Tweak error and warning colors and fix `StatusWarning` icon visibility on light themes

rsubtil opened the pull request:
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Fixes #87987

Originally this PR focused on fixing visibility of the `StatusWarning`, but after successive iterations and positive feedback, a few more relevant alterations were made:

- Removed `StatusWarning` and `StatusError` icons from the exclusion list to be able to convert its colors
- Set `error_color` to the same color used in `Node3D` (`#BF5A50` -> `#CD3838`)
- Shift `warning_color` to a more orange saturated hue (`#A69042` -> `#D18F19`)
- Tweak `EditorLogFilterButton` appearance (more comparison screenshots at https://github.com/godotengine/godot/pull/88058#issuecomment-1935833345):
    - Increased icon `normal` color opacity from 40% to 80%
    - Removed icon `hover` color modulate, making it now the original color, *aka* 100%
    - Lightened gray brackground on `pressed` state by 50%

---

<details><summary><b>Comparison screenshots</b></summary>

| Before | After |
| - | - |
|![1_old](https://github.com/godotengine/godot/assets/6501975/2ad2b80e-bfeb-4408-bb2a-8e48837ab113) |![1_new](https://github.com/godotengine/godot/assets/6501975/bba4d9c8-609b-4fb6-9252-db8c486a7078) |
|![2_old](https://github.com/godotengine/godot/assets/6501975/406c7c6b-4d68-4afd-90c4-889f9ae0ce6b) |![2_new](https://github.com/godotengine/godot/assets/6501975/b4bf4c6b-a561-473b-be44-a621b3a0ec2d) |
|![3_old](https://github.com/godotengine/godot/assets/6501975/72afa890-ad5b-4e21-907b-6c5e16714913) |![3_new](https://github.com/godotengine/godot/assets/6501975/e5e0ecfd-9891-4bb0-b917-ae760037e27e) |
|![4_old](https://github.com/godotengine/godot/assets/6501975/47f2ff07-b639-4b1c-8f2c-7a2a57464198) |![4_new](https://github.com/godotengine/godot/assets/6501975/24c28a0a-1cce-49c7-b8a5-d071a7a755c8) |
|![5_old](https://github.com/godotengine/godot/assets/6501975/d8a87965-5672-4a0a-b586-995b8c4c86b0) |![5_new](https://github.com/godotengine/godot/assets/6501975/3f640cb4-0e3b-435d-8f85-81b33a77b683) |
|![6_old](https://github.com/godotengine/godot/assets/6501975/a82a1f5e-3bd7-4c61-ac76-3a2e888ccaeb) |![6_new](https://github.com/godotengine/godot/assets/6501975/bb4767e7-c3d7-4d8b-a8d8-cad7a5e94a93) |
|![7_old](https://github.com/godotengine/godot/assets/6501975/ba13f2b6-fc7a-4b81-a375-9641f1ba83c9) |![7_new](https://github.com/godotengine/godot/assets/6501975/8cb453b0-5fe3-420b-8539-57fceaad17c4) |

</details>

rsubtil committed:
Fix warning icon visibility on light themes

adamscott reviewed (approved):
As simple as that! Thanks @rsubtil !

akien-mga commented:
On the second "After" screenshot, the error color also seems to lack contrast when the button is not toggled.

The toggle color itself seems to create contrast issues IMO. So I think we might need a bit more work than just special casing StatusWarning.

Calinou commented:
> The toggle color itself seems to create contrast issues IMO. So I think we might need a bit more work than just special casing StatusWarning.

The icon's opacity is decreased when the button isn't pressed – we probably need to lessen this effect, especially on light theme.

akien-mga commented:
> (I've also included the hover state as I noticed it looked bad for the yellow icon)

Yeah I was about to point out that this is still lacking contrast. Overall the colors for both StatusError and StatusWarning on the light theme seem too washed out / desaturated IMO.

And the pressed button color (blue-grayish) also seems too dark / dull IMO, at least from close up, maybe in context it fills more appropriate.

rsubtil commented:
> Yeah I was about to point out that this is still lacking contrast. Overall the colors for both StatusError and StatusWarning on the light theme seem too washed out / desaturated IMO.

- For `StatusError` we could make a color conversion for the same color being used for `Node3D`, which is a more saturated red.

- For `StatusWarning`, the color is already being converted. Increasing its saturation doesn't work, as yellow will always be problematic on white, and it affects other elements adversely. But, I liked the effect when I shifted to a more "orange" hue.

> And the pressed button color (blue-grayish) also seems too dark / dull IMO, at least from close up, maybe in context it feels more appropriate.

The effect is indeed much less pronounced on the dark theme, likely because of the already dark blue tint overall:
| Dark | Light |
| - | - |
| ![image](https://github.com/godotengine/godot/assets/6501975/75930335-042c-4d2a-9010-55ee0af29b47) | ![image](https://github.com/godotengine/godot/assets/6501975/36849b19-2661-41ba-ab4d-dabc9c759656) |

I think reducing greatly the grayish tint works well for light themes, as the blue border is already a clear indicator of the button being toggled.

---

All in all, the changes are now as follows:

- Removed `StatusError` from the exclusion list to convert its colors
- Set `error_color` to the same color used in `Node3D` (`#BF5A50` -> `#CD3838`)
- Set `warning_color` to a more saturated color, and with a more orange tint (`#A69042` -> `#D18F19`)
- (On the output filter buttons only)
    - Increased icon `normal` color opacity from 40% to 80%
    - Removed icon `hover` color modulate (now original color, aka 100%)
    - Lightened gray brackground on `pressed` state by 50%


| | Normal | Hover | Pressed |
|-|-|-|-|
|Before|![image](https://github.com/godotengine/godot/assets/6501975/e901b30e-91fb-443c-8730-7fe2b3e87dc5) |![image](https://github.com/godotengine/godot/assets/6501975/7aed27d0-9b76-4736-989c-548b3399991a) |![image](https://github.com/godotengine/godot/assets/6501975/1fb71c70-9035-415d-b0fb-94fdb7cd4b0e) |
|[1st iteration](https://github.com/godotengine/godot/pull/88058#issuecomment-1934070071)| ![image](https://github.com/godotengine/godot/assets/6501975/db0715e5-76c8-4851-89a9-044a0af039ac) | ![image](https://github.com/godotengine/godot/assets/6501975/9b6b036f-0958-4763-875a-6cd29c31b8da) | ![image](https://github.com/godotengine/godot/assets/6501975/9037db0b-d293-4096-b07f-b77e8ae2c30d) |
|2nd iteration| ![image](https://github.com/godotengine/godot/assets/6501975/2b6e2c69-dfdf-4b98-959a-776f27fb8330)|![image](https://github.com/godotengine/godot/assets/6501975/5e5462a8-155d-4253-aa4f-d77e521e1a91) |![image](https://github.com/godotengine/godot/assets/6501975/3ac37c90-1ef5-41d0-bc2c-732f050b31ae)|

>[!WARNING]
>This changes some core colors used throughout all the editor. The `error_color` is subtle, but the `warning_color` is more noticeable:
>
>| Before | After |
>|-|-|
>|![image](https://github.com/godotengine/godot/assets/6501975/74ee9e5e-c207-4e9a-970b-6d76a7fd0b07)|![image](https://github.com/godotengine/godot/assets/6501975/2bf5ff96-842d-44ff-a9c7-3c5f7f7f959e)|

Mickeon commented:
I dig the orange hue for the warning color in the light theme. It's more contrasted.
I am not sure about a few other changes that _also_ affect the default theme, unless extremely small, as they may affect more places that one may think at a glance. Needs some testing.

Calinou commented:
The new warning color is great! I've always found the previous warning color less than ideal on a light theme, so this is a big improvement.

Calinou commented on editor/themes/editor_theme_manager.cpp:
```diff
@@ -396,8 +396,8 @@ void EditorThemeManager::_create_shared_styles(const Ref<EditorTheme> &p_theme,
 			p_config.success_color = p_config.success_color.lerp(p_config.mono_color, 0.35);
-			p_config.warning_color = p_config.warning_color.lerp(p_config.mono_color, 0.35);
-			p_config.error_color = p_config.error_color.lerp(p_config.mono_color, 0.25);
+			p_config.warning_color = Color("#d18f19");
+			p_config.error_color = Color("#cd3838");
```
I suggest using the `Color(float, float, float)` form which avoids storing a string in the binary, and therefore reduces binary size ever so slightly. I did this on all colors in the codebase a while ago and it reduced binary size by a non-negligible amount.

`Color::hex(0xrrggbbaa)` might reduce binary size further by using 1 integer instead of 3 floats, but that should be evaluated later.

KoBeWi replied:
`Color::hex(0xrrggbbaa)` can be shortened to `Color(0xrrggbbaa)`, no? It's a direct replacement for hex string.

Mickeon replied:
My assumption was that the compiler was smarter than this but I suppose that these are among the many things to keep in mind. I did not expect for the Colors to be basically generated on the spot even for constant Strings.

rsubtil replied:
I went with the 3-float approach then, but I agree reducing it in the future to a single hex int sounds interesting, especially for iterating on the colors and not having to constantly convert them to a [0, 1] float range.

Calinou replied:
> By the way, I may be wrong but I believe the bits are reversed in pure hex: 0xaabbggrr.

`ARGB` was used in Godot 3.x, but Godot 4.0 switched to `RGBA`.

rsubtil commented:
Since the new changes ended up affecting the overall interface significantly, and there was very positive feedback on the warning color changes, I've edited this PR description to better describe the changes done, along with some new comparison screenshots.

akien-mga reviewed (approved):
Looks good to me.

akien-mga commented:
Thanks!
